### ISSUE TITLE
Wake delay and BSOD

### ISSUE THREAD

**Issue by ghost** [TheWidlarzGroup/react-native-video#44]
Locking the device while the video element is on-screen causes up to a 3-4 second delay when the phone attempts to wake, and causes the video component to turn black.

**Affected example app:** https://github.com/jhabdas/lumpen-radio/tree/v1.0.0-rc

To replicate follow the instructions for setting up the app. Start the app on an actual device, unplug the device from USB, turn the screen off with the app running and then unlock the device after 20 seconds.

When the screen does eventually come back to life (usually after a few home button presses) the video disappears and the video content area goes black.

**Comment by ghost** [comment]
Was able to fix this by simply adding a listener using [`AppStateIOS`](https://facebook.github.io/react-native/docs/appstateios.html) module and setting the `repeat` prop to `false` when the app was _not_ `active`. Closing this issue as it's really more of an enhancement. Feel free to reopen as such.

**Comment by ghost** [comment]
Quick update. Still seeing this issue sometimes even with use of AppStateIOS. Haven't figured out root cause as it's difficult to reproduce consistently. What I can mention is that, when the device wakes, it's often running hot and the video _seems_ to appear momentarily before getting replaced with a black background.

In closed beta right now and none of my internal testers have brought it up yet, but if they do I may have to get more surgical to address.

**Comment by ghost** [comment]
Thanks for re-opening @brentvatne. Just had my first beta user report the issue and mentioned the black screen. Said he'd left the app running in the background and when he came back to it some time later this issue occurred.

**Comment by brentvatne** [comment]
@jhabdas - I think the desired behaviour here is to automatically pause the video when we background the app, and resume it when you jump back in. But we can make this controllable via a prop as well, because there may be situations where you want to have audio/video running in background (eg: a media player app like Rdio or Spotify) :smile:

**Comment by ghost** [comment]
Sounds good. Will continue with your recommendations and tightening up my state management and update later. Hope to see some sweet control center integration with the audio going into this awesome component. Keep up the great work!

**Comment by ghost** [comment]
Based on additional research and debugging today I was able to replicate this issue even without the video looping. As such I have adjusted the title and description of this issue, cut a [pre-release version of the affected app](https://github.com/jhabdas/lumpen-radio/tree/v1.0.0-rc) and updated the steps to replicate.

As far as I can tell, this issue should be able to be replicated on any app which uses this component if the device is locked (replicated consistently on iPhone 6) without first sending the app into the background, and with the USB connection to Xcode unplugged.

A workaround for this issue would be to remove the video component when the app state goes into the `background`, but, from what I can see, the [AppStateIOS](https://github.com/facebook/react-native/blob/v0.6.0/React/Modules/RCTAppState.m) module doesn't fire the `UIApplicationDidEnterBackgroundNotification` event in time for it to make it over the bridge; the JS event listeners can't catch it in time to respond.

I tried to speed things up a little by dispatching my own [`inactive`](https://github.com/facebook/react-native/issues/524) event across the bridge with some light abuse of `appStateDidChange`, though, unfortunately, even this event wasn't fast enough and was only received after the app was unlocked (as if it were somehow deferred).

Here's some code for anyone who want's to give the `inactive` hack a try:

``` objective-c
- (void) applicationWillResign
{
  [self.bridge.eventDispatcher sendDeviceEventWithName:@"appStateDidChange"
                                                  body:@{@"app_state": @"inactive"}]; // received on JS side of bridge only after the device is unlocked
  NSLog(@"derp"); // outputs to log a moment after the screen is locked
}

- (void) subscribeToNotifications
{
  [[NSNotificationCenter defaultCenter]
   addObserver:self
   selector:@selector(applicationWillResign)
   name:UIApplicationWillResignActiveNotification
   object:nil];
}
```

Unfortunately this means I'm going to have to remove this component from my app until a suitable workaround or fix can be found. I'd love to study any app using this component which does not express this issue if someone can point me in the right direction.

**Comment by brentvatne** [comment]
@jhabdas - I can't seem to get your project running :( followed your README instructions and got this: 
![screen shot 2015-06-25 at 3 25 22 pm](https://cloud.githubusercontent.com/assets/90494/8366922/6eb5d4e8-1b4e-11e5-97a9-f41a59492c10.png)

I tried the steps you mentioned above on an app that I have which uses video and I didn't have the problem, but I tried a fix anyways and pushed it in `0.4.8` - can you let me know if that works for you? (https://github.com/brentvatne/react-native-video/commit/1f901be3638d7c6c4ec8599ddd834c7d43837fb4)

**Comment by jhabdas** [comment]
Thanks!! Will do first thing mañana. Those Pod files can be tricky.

**Comment by ghost** [comment]
Thanks for cutting a new version @brentvatne. Unfortunately I'm still seeing the same problem after pulling in `0.4.8`. Because my app's about 2 minors behind `react-native` now I'm going to pull it forward to the latest stable to see if that helps clear things up. I also subbed #65, though I'm not sure it's necessary as, at least in my case, I noticed the problem occurred with and without looping enabled.

I'll update with more info mid to late next week on the outcomes of the update. Meanwhile, unless others are seeing this issue, it may be pertinent to bump again to pull the changes in https://github.com/brentvatne/react-native-video/commit/1f901be3638d7c6c4ec8599ddd834c7d43837fb4. Your call, of course.

Thanks for your time and attention on this. More info to follow.

**Comment by brentvatne** [comment]
Hey @jhabdas - that's very strange, I wonder why this would be happening if it's being paused when the app becomes inactive. I haven't seen any mention of this in googling for AVPlayer issues so it doesn't seem to be specific to that. @johanneslumpe - any ideas?

**Comment by ghost** [comment]
Resolved with bump to React Native `v0.5.0` (or `v0.6.0`) and React Native Video `v0.4.9`. Looks like there were some dispatcher bugs resolved in [`v0.5.0-rc`](https://github.com/facebook/react-native/releases/tag/v0.5.0-rc) which I overlooked while debugging, and when I test drove the latest version of this component the other day I did so using a `0.4.x` version of React Native. Derp.

**Comment by brentvatne** [comment]
Awesome! :smile:
